control-plane/subcommand/partition-init/command.go

```diff
@@ -0,0 +1,208 @@
+func (c *Command) Run(args []string) int {
+	c.once.Do(c.init)
+	if err := c.flags.Parse(args); err != nil {
+		return 1
+	}
+	if len(c.flags.Args()) > 0 {
+		c.UI.Error("Should have no non-flag arguments.")
+		return 1
+	}
+
+	// Validate flags
+	if err := c.validateFlags(); err != nil {
+		c.UI.Error(err.Error())
+		return 1
+	}
+	var cancel context.CancelFunc
+	c.cmdTimeout, cancel = context.WithTimeout(context.Background(), c.flagTimeout)
+	// The context will only ever be intentionally ended by the timeout.
+	defer cancel()
+
+	var err error
+	c.log, err = common.Logger(c.flagLogLevel, c.flagLogJSON)
+	if err != nil {
+		c.UI.Error(err.Error())
+		return 1
+	}
+
+	serverAddresses := c.flagServerAddresses
+	// Check if the provided addresses contain a cloud-auto join string.
+	// If yes, call godiscover to discover addresses of the Consul servers.
+	if len(c.flagServerAddresses) == 1 && strings.Contains(c.flagServerAddresses[0], "provider=") {
+		var err error
+		serverAddresses, err = godiscover.ConsulServerAddresses(c.flagServerAddresses[0], c.providers, c.log)
+		if err != nil {
+			c.UI.Error(fmt.Sprintf("Unable to discover any Consul addresses from %q: %s", c.flagServerAddresses[0], err))
+			return 1
+		}
+	}
+
+	scheme := "http"
+	if c.flagUseHTTPS {
+		scheme = "https"
+	}
+
+	serverAddr := fmt.Sprintf("%s:%d", serverAddresses[0], c.flagServerPort)
+	consulClient, err := consul.NewClient(&api.Config{
+		Address: serverAddr,
+		Scheme:  scheme,
+		Token:   "", // TODO: Figure out if a token is required here and which one that is.
+		TLSConfig: api.TLSConfig{
+			Address: c.flagConsulTLSServerName,
+			CAFile:  c.flagConsulCACert,
+		},
+	})
+	if err != nil {
+		c.log.Error(fmt.Sprintf("Error creating Consul client for addr %q: %s", serverAddr, err))
+		return 1
+	}
+
+	partition, _, err := consulClient.Partitions().Read(c.cmdTimeout, c.flagPartitionName, &api.QueryOptions{})
+	// The API does not return an error if the Partition does not exist. It returns a nil Partition.
+	if partition == nil {
+		for {
+			// Retry Admin Partition creation until it succeeds, or we reach the command timeout.
+			_, _, err = consulClient.Partitions().Create(c.cmdTimeout, &api.AdminPartition{Name: c.flagPartitionName}, &api.WriteOptions{})
+			if err == nil {
+				c.log.Info(fmt.Sprintf("Successfully created Admin Partition '%s'", c.flagPartitionName))
+				return 0
+			}
+			c.log.Error(fmt.Sprintf("Error creating partition '%s': '%s'", c.flagPartitionName, err))
+			c.log.Info("Retrying in " + c.retryDuration.String())
+			// Wait on either the retry duration (in which case we continue) or the
+			// overall command timeout.
+			select {
+			case <-time.After(c.retryDuration):
+				continue
+			case <-c.cmdTimeout.Done():
+				return 1
+			}
+		}
+	} else if err != nil {
+		c.log.Error(fmt.Sprintf("Error reading Partition '%s' from Consul: %s", c.flagPartitionName, err))
+		return 1
+	}
```

Comment:
thinking: is there a use-case where we want it to retry even if at first it can't talk to consul for some reason? Like what if the consul server isn't up yet for some reason?

Reply:
makes sense